design/ozone.md

```diff
@@ -0,0 +1,246 @@
+This transformation can be applied to any of the k8s resources.
+
+Until now it is very similar to Kustomize, but here we have more predefined transformations (DaemonSet to StatefulSet converter, NodePort service generator, value replacer).
+
+Even better the transformations can be reused. The previous definition can be reused by putting it to the right directory:
+
+```yaml
+name: ozone/prometheus
+description: Enable prometheus monitoring in Ozone
+---
+- type: Add
+  path: ["spec","template","spec","metadata","annotations"]
+  trigger:
+     metadata:
+        name: datanode
+  value:
+     prometheus.io/scrape: "true"
+     prometheus.io/port: "9882"
+     prometheus.io/path: "/prom"
+```
+
+After this step, you can use it as
+
+```yaml
+type: ozone/prometheus
+```
+
+Or from golang API: 
+
+```golang
+
+context, err := Initialize("../testdata/api")
+if err != nil {
+   return err
+}
+
+err = context.AppendCustomProcessor("ozone/prometheus", make(map[string]string))
+if err != nil {
+   return err
+}
+
+context.Render()
+
+s, err := context.Resources()[0].Content.ToString()
+if err != nil {
+   return err
+}
+
+assert.Contains(t, s, "prometheus.io/scrape: \"true\"")
+```
+
+### Create k8s resource from operator
+
+One easy way to integrate the exising resource generation of Ozone is to use it from golang API. This is similar to the approach of [ceph csi](https://github.com/rook/rook/tree/master/pkg/operator/ceph/csi) which reads the k8s template files from the disk.
+
+The main diference here is that before the final adjustment we can reuse the basic transformations which are already implemented by ozone.
+
+At the end of the transformations we will have a list of real K8s resources object, where we can do any final adjustment on the API object and send them to the k8s API services.
+
+The main logic is the following: 
+
+ * The resource generator can be initialized with the main k8s definition directory (which is included in the Ozone release, and included in the rook operator image)
+ * The transformations can be added programatically based on CRD properties
+ * The real k8s golang objects are generated
+ * After final adjustment in golang, we can post all the objects to the K8S api
+
+### CRD
+
+A good operator CRD provides a good balance between the simplicity and power. By default the CRD should be as simple as possible but we should keep the possibility to customize any segment of the generated cluster.
+
+This can be achieved by the use k8s generator tool.
+
+ 1. The default cluster will be generated based on the main attributes of CRD
+ 2. Additional attributes can trigger additional transformations
+ 3. Finally we can provide an option to define any custom transformations
+
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: rook-ozone
+---
+apiVersion: ozone.rook.io/v1alpha1
+kind: ObjectStore
+metadata:
+  name: my-store
+  namespace: rook-ozone
+spec:
+  components: #high lebel options to turn off/on component groups
```

Comment:
You will probably have settings that you want to add to the different types of components. To support adding settings by component, you might consider this structure:
```yaml
spec:
  csi:
    deploy: true
  recon: 
    deploy: true
  s3g:
    deploy: true
```
The settings look unnecessary when it is this simple, but it will give you flexibility for the future.

Reply:
Fair enough. I updated it. I would use `enabled` instead of `deploy` as it seems to be used by other components (eg. ceph dashboard) but following the suggested approach:

```
spec:
  components: #component specific settings
     csi:
       enable: true
     recon:
       enable: true
     s3g: 
       enable: true
```

Reply:
Update: I added the persistence fields to here as well (based on a comment from below)